rust/lance-encoding/src/encoder.rs

```diff
@@ -272,26 +270,37 @@ impl CoreArrayEncodingStrategy {
     }
 }
 
+fn get_dict_encoding_threshold() -> u64 {
+    env::var("LANCE_DICT_ENCODING_THRESHOLD")
+        .ok()
+        .and_then(|val| val.parse().ok())
+        .unwrap_or(100)
+}
+
 // check whether we want to use dictionary encoding or not
 // by applying a threshold on cardinality
-// returns true if cardinality < min(100, num_total_rows)
+// returns true if cardinality < threshold but false if the total number of rows is less than the threshold
 // The choice to use 100 is just a heuristic for now
-fn check_dict_encoding(arrays: &[ArrayRef]) -> bool {
-    let mut unique_values = HashSet::new();
-
+// hyperloglog is used for cardinality estimation
+// error rate = 1.04 / sqrt(2^p), where p is the precision
+// and error rate is 1.04 / sqrt(2^12) = 1.56%
+fn check_dict_encoding(arrays: &[ArrayRef], threshold: u64) -> bool {
     let num_total_rows = arrays.iter().map(|arr| arr.len()).sum::<usize>();
-    let thresh = min(100, num_total_rows);
+    if num_total_rows < threshold as usize {
+        return false;
+    }
+    const PRECISION: u8 = 12;
+
+    let mut hll: HyperLogLogPlus<String, RandomState> =
+        HyperLogLogPlus::new(PRECISION, RandomState::new()).unwrap();
 
-    // Checking cardinality for elements in all arrays together.
-    for arr in arrays.iter() {
+    for arr in arrays {
         let string_array = arrow_array::cast::as_string_array(arr);
-        let arr_len = string_array.len();
-        for i in 0..arr_len {
-            if !string_array.is_null(i) {
-                unique_values.insert(string_array.value(i));
-                if unique_values.len() >= thresh {
-                    return false;
-                }
+        for value in string_array.iter().flatten() {
+            hll.insert(value);
+            let estimated_cardinality = hll.count() as u64;
```

Comment:
I wonder if `count` is still cheap enough to justify using as shortcut here?  Maybe only call `count` every 100 passes through the loop?

Not important right now.  We can micro-optimize later with benchmarks.

Reply:
Good catch. For hyperloglog, the `count` operation is an `O(1)` operation, however, I am not sure if the constant factor matters in this case.

Reply:
I’m considering submitting another PR to include the following:
* More accurately estimating the benefits of encoding by calculating the actual savings: https://github.com/lancedb/lance/pull/2555#discussion_r1668589005
* Calling count every 100 passes through the loop

Would this be helpful, or should we wait for more detailed benchmark results?